sockeye/encoder.py

```diff
@@ -413,3 +470,173 @@ def get_rnn_cells(self) -> List[mx.rnn.BaseRNNCell]:
+        self.num_embed = num_embed
+        self.max_filter_width = max_filter_width
+        self.num_filters = num_filters[:]
+        self.pool_stride = pool_stride
+        self.num_highway_layers = num_highway_layers
+        self.prefix = prefix
+        self.dropout = dropout
+        self.prefix = prefix
+
+        self.conv_weight = {filter_width: mx.sym.Variable("%s%s%d%s" % (self.prefix, "conv_", filter_width, "_weight"))
+                            for filter_width in range(1, self.max_filter_width + 1)}
+        self.conv_bias = {filter_width: mx.sym.Variable("%s%s%d%s" % (self.prefix, "conv_", filter_width, "_bias"))
+                          for filter_width in range(1, self.max_filter_width + 1)}
+
+        self.gate_weight = [mx.sym.Variable("%s%s%d%s" % (self.prefix, "gate_", i, "_weight"))
+                            for i in range(self.num_highway_layers)]
+        self.gate_bias = [mx.sym.Variable("%s%s%d%s" % (self.prefix, "gate_", i, "_bias"))
+                          for i in range(self.num_highway_layers)]
+
+        self.transform_weight = [mx.sym.Variable("%s%s%d%s" % (self.prefix, "transform_", i, "_weight"))
+                                 for i in range(self.num_highway_layers)]
+        self.transform_bias = [mx.sym.Variable("%s%s%d%s" % (self.prefix, "transform_", i, "_bias"))
+                               for i in range(self.num_highway_layers)]
+
+    def encode(self, data: mx.sym.Symbol, data_length: mx.sym.Symbol, seq_len: int) -> mx.sym.Symbol:
+        """
+        Encodes data given sequence lengths of individual examples and maximum sequence length.
+
+        :param data: Input data (batch_size, seq_len, num_embed).
+        :param data_length: Vector with sequence lengths.
+        :param seq_len: Maximum sequence length.
+        :return: Encoded input data.
+        """
+
+        total_num_filters = sum(self.num_filters)
+        encoded_seq_len = self.get_encoded_seq_len(seq_len)
+
+        # (batch_size, channel=1, seq_len, num_embed)
+        data = mx.sym.Reshape(data=data, shape=(-1, 1, seq_len, self.num_embed))
+
+        # Convolution filters of width 1..N
+        conv_outputs = []
+        for filter_width, num_filter in enumerate(self.num_filters, 1):
+            # "half" padding: output length == input length
+            pad_before = ceil((filter_width - 1) / 2)
+            pad_after = floor((filter_width - 1) / 2)
+            # (batch_size, channel=1, seq_len + (filter_width - 1), num_embed)
+            padded = mx.sym.pad(data=data,
+                                mode="constant",
+                                constant_value=0,
+                                pad_width=(0, 0, 0, 0, pad_before, pad_after, 0, 0))
+            # (batch_size, num_filter, seq_len, num_scores=1)
+            conv = mx.sym.Convolution(data=padded,
+                                      #cudnn_tune="off",
```

Comment:
There are some memory/speed tuning options here.  This is one of many performance optimization TODOs (commented out -> pick what cuDNN thinks is best).